### PR TITLE
Enabling multi indexing

### PR DESCRIPTION
**Pull request by StrikingThirteen**
## Overview

Description:
This commit enables multi value indexing by enabling CorfuTable of having indexes with mono index function or multi-index function. In other words, when a {K, V} being indexed by a multi index function, it will emit multiple index mappings.

Why should this be merged: 
This features is going to be used by clients (e.g. management plane) to emit multiple index mappings for each key/value pair.

Related issue(s) (if applicable): #1285 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: test/src/test/java/org/corfudb/integration/ServerRestartIT.java [hunk 1]
```diff
@@ -733,6 +734,64 @@ public void testCorfuTableIndexReconstruction() throws Exception {
         // Stop the corfu server.
         assertThat(shutdownCorfuServer(corfuProcess)).isTrue();
     }
-}
 
+    @Test
+    public void testCorfuTableMultiIndexReconstruction() throws Exception {
```
Comment (medhavidhawan) [inline test/src/test/java/org/corfudb/integration/ServerRestartIT.java]: good test. Please add comments at the top. It helps build context faster when we come back to these things in future.

Comment (StrikingThirteen) [reply test/src/test/java/org/corfudb/integration/ServerRestartIT.java]: Done.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java [hunk 2]
```diff
@@ -144,10 +175,9 @@ public int hashCode() {
          * Obtain the {@link CorfuTable.IndexFunction} via its registered {@link CorfuTable.IndexName}.
          *
          * @param name name of the {@code IndexKey} previously registered.
-         * @param <I>  type of the {@code IndexKey}.
          * @return the instance of {@link CorfuTable.IndexFunction} registered to the lookup name.
          */
-        <I extends Comparable<?>> Optional<CorfuTable.IndexFunction<K, V, I>> get(CorfuTable.IndexName name);
+        Optional<CorfuTable.Index<K, V, ? extends Comparable<?>>> get(CorfuTable.IndexName name);
```
Comment (corfu-repo) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java]: CRITICAL Remove usage of generic wildcard type. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1452)

**Comment by corfu-repo** [comment]
SonarQube analysis reported 7 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![CRITICAL][CRITICAL] 3 critical
* ![MAJOR][MAJOR] 2 major
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![BLOCKER][BLOCKER] [CorfuTable.java#L189](https://github.com/CorfuDB/CorfuDB/blob/f7d4adf0279f132dc316557c61263a975f254106/runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java#L189): Rename method "empty" to prevent any misunderstanding/clash with field "EMPTY" defined on line 162. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1845)
1. ![CRITICAL][CRITICAL] [CorfuTable.java#L159](https://github.com/CorfuDB/CorfuDB/blob/f7d4adf0279f132dc316557c61263a975f254106/runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java#L159): Move constants to a class or enum. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1214)
1. ![CRITICAL][CRITICAL] [CorfuTable.java#L748](https://github.com/CorfuDB/CorfuDB/blob/f7d4adf0279f132dc316557c61263a975f254106/runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java#L748): Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [ServerRestartIT.java#L176](https://github.com/CorfuDB/CorfuDB/blob/f7d4adf0279f132dc316557c61263a975f254106/test/src/test/java/org/corfudb/integration/ServerRestartIT.java#L176): Remove this use of "Thread.sleep()". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)
1. ![MAJOR][MAJOR] [ServerRestartIT.java#L289](https://github.com/CorfuDB/CorfuDB/blob/f7d4adf0279f132dc316557c61263a975f254106/test/src/test/java/org/corfudb/integration/ServerRestartIT.java#L289): Remove this use of "Thread.sleep()". [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2925)
1. ![INFO][INFO] [CorfuTable.java#L559](https://github.com/CorfuDB/CorfuDB/blob/f7d4adf0279f132dc316557c61263a975f254106/runtime/src/main/java/org/corfudb/runtime/collections/CorfuTable.java#L559): Complete the task associated to this TODO comment. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)

[BLOCKER]: https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER'
[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**